vulnerabilities/data_source.py

```diff
@@ -0,0 +1,371 @@
+from typing import Sequence
+from typing import Set
+from typing import Tuple
+
+import pygit2
+from packageurl import PackageURL
+
+
+@dataclasses.dataclass
+class Advisory:
+    """
+    This data class expresses the contract between data sources and the import runner.
+    Data sources are expected to be usable as context managers and generators, yielding batches of Advisory sequences.
+
+    NB: There are two representations for package URLs that are commonly used by code consuming this data class;
+        PackageURL objects and strings. As a convention, the former is referred to in variable names, etc. as
+        "package_urls" and the latter as "purls".
+    """
+    summary: str
+    impacted_package_urls: Sequence[PackageURL]
+    resolved_package_urls: Sequence[PackageURL] = dataclasses.field(default_factory=list)
+    references: Sequence[str] = dataclasses.field(default_factory=list)
+    cve_id: Optional[str] = None
+
+    @property
+    def impacted_purls(self) -> Set[str]:
+        return {str(p) for p in self.impacted_package_urls}
+
+    @property
+    def resolved_purls(self) -> Set[str]:
+        return {str(p) for p in self.resolved_package_urls}
+
+
+class InvalidConfigurationError(Exception):
+    pass
+
+
+@dataclasses.dataclass
+class DataSourceConfiguration:
+    batch_size: int
+
+
+class DataSource(ContextManager):
+    """
+    This class defines how importers consume advisories from a data source.
+
+    It makes a distinction between newly added records since the last run and modified records. This allows the import
+    logic to pick appropriate database operations.
+    """
+
+    CONFIG_CLASS = DataSourceConfiguration
+
+    def __init__(
+            self,
+            batch_size: int,
+            last_run_date: Optional[datetime] = None,
+            cutoff_date: Optional[datetime] = None,
+            config: Optional[Mapping[str, Any]] = None,
+    ):
+        """
+        Create a DataSource instance.
+
+        :param batch_size: Maximum number of records to return from added_advisories() and updated_advisories()
+        :param last_run_date: Optional timestamp when this data source was last inspected
+        :param cutoff_date: Optional timestamp, records older than this will be ignored
+        :param config: Optional dictionary with subclass-specific configuration
+        """
+        config = config or {}
+        try:
+            self.config = self.__class__.CONFIG_CLASS(batch_size, **config)
+            # These really should be declared in DataSourceConfiguration above but that would prevent DataSource
+            # subclasses from declaring mandatory parameters (i.e. positional arguments)
+            setattr(self.config, 'last_run_date', last_run_date)
+            setattr(self.config, 'cutoff_date', cutoff_date)
+        except Exception as e:
+            raise InvalidConfigurationError(str(e))
+
+        self.validate_configuration()
+
+    def __enter__(self):
```

Comment:
nit: `__enter__ ` and `__exit__` are already defined in the `ContextManager` abstract class, which is inherited by this class(DataSource) . These can be safely removed.

Reply:
`ContextManager` is basically https://github.com/python/cpython/blob/eff870b618ca6f6b7a60a271f15af7e54b8a1b97/Lib/contextlib.py#L15 as mentioned in https://docs.python.org/3/library/typing.html#typing.ContextManager

Reply:
> nit: `__enter__ ` and `__exit__` are already defined in the `ContextManager` abstract class, which is inherited by this class(DataSource) . These can be safely removed.

Right. First I had some code in there but then moved it elsewhere. I'll remove it.

Reply:
Turns out if I remove them, some tests fail because inheriting from `DataSource`, which inherits from `ContextManager` does not give `MockDataSource` these methods. ¯\_(ツ)_/¯